tests/functional/test_build_constraints.py

```diff
@@ -0,0 +1,176 @@
+"""Tests for the build constraints feature."""
+
+from __future__ import annotations
+
+from pathlib import Path
+
+import pytest
+
+from tests.lib import PipTestEnvironment, TestPipResult, create_test_package_with_setup
+
+
+def _create_simple_test_package(script: PipTestEnvironment, name: str) -> Path:
+    """Create a simple test package with minimal setup."""
+    return create_test_package_with_setup(
+        script,
+        name=name,
+        version="1.0",
+        py_modules=[name],
+    )
+
+
+def _create_constraints_file(
+    script: PipTestEnvironment, filename: str, content: str
+) -> Path:
+    """Create a constraints file with the given content."""
+    constraints_file = script.scratch_path / filename
+    constraints_file.write_text(content)
+    return constraints_file
+
+
+def _run_pip_install_with_build_constraints(
+    script: PipTestEnvironment,
+    project_dir: Path,
+    build_constraints_file: Path,
+    extra_args: list[str] | None = None,
+    expect_error: bool = False,
+) -> TestPipResult:
+    """Run pip install with build constraints and common arguments."""
+    args = [
+        "install",
+        "--no-cache-dir",
+        "--build-constraint",
+        str(build_constraints_file),
+        "--use-feature",
+        "build-constraint",
+    ]
+
+    if extra_args:
+        args.extend(extra_args)
+
+    args.append(str(project_dir))
+
+    return script.pip(*args, expect_error=expect_error)
+
+
+def _assert_successful_installation(result: TestPipResult, package_name: str) -> None:
+    """Assert that the package was successfully installed."""
+    assert f"Successfully installed {package_name}" in result.stdout
```

Comment:
Isn't there an `result.assert_installed()` helper for this? It does require a version though.